documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,888 @@
+
+Although the encap information will still be saved in the metadata bus, however, for stateless decap, we will not create the encap for the reverse flow. Otherwise (for the stateful decap), we will create the encap for the reverse flow by reversing the source and destination information.
+
+```json
+{
+    "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "stateless": false },
+    "DASH_SAI_VNI_TABLE|12346": { "direction": "outbound", "stateless": true }
+}
+```
+
+#### 5.5.3. Encap fields handling
+
+##### 5.5.3.1. Handling DSCP
+
+The DASH pipeline provides 2 modes for handling the DSCP: "Preserve model" and "Pipe model".
+
+Preserve model works very similar to `SAI_TUNNEL_DSCP_MODE_UNIFORM_MODEL` with a slight difference:
+
+- The DSCP value of outer most encap header will be preserved.
+- When any encap being added, the preserved DSCP value will be copied to the new encap header.
+- The DSCP value of overlay packet (customer packet) shall never be modified or preserved and copied to the outer header.
+
+On the other hand, pipe model works very similar to `SAI_TUNNEL_DSCP_MODE_PIPE_MODEL`:
+
+- The DSCP value of existing encaps will not be preserved.
+- When any encap being added, the DASH pipeline will specify the DSCP value, and all routing action will take this value and copy it to the encaps.
+
+If no encaps are added, say, for traffic sending to Internet, the DSCP value of overlay packet will be exposed as it is, even-if original packet arrives with a non-zero value.
+
+This gives the cloud infra full control over the DSCP value and will prevent the customer from spoofing the fields.
+
+##### 5.5.3.2. Handling TTL
+
+TTL behavior for encap shall be "pipe" model (similar to SAI_TUNNEL_TTL_MODE_PIPE_MODEL):
+
+- When adding encaps, TTL value shall be default set to 64.
+- DASH pipeline shall not modify the TTL values in the overlay packet (customer packet).
+
+#### 5.5.4. Encap preservation
+
+Sometimes, depends on the scenario to implement, the customer might want to preserve certain original encaps in the outgoing traffic. For example, say we receive a packet with structure: overlay -> underlay0 -> underlay1 -> underlay2. And we want to remove or update underlay0, preserve underlay1 and remove underlay2. This gives us the problem of handling all the CRUD combinations of all encaps, including structure changes: after removing underlay0, should underlay1 becomes underlay0 or should we keep it as underlay1? All these things affects the encap related routing actions and final packet we create.
+
+Since all the encap information is preserved in the metadata bus for flow creation anyway, to solve this problem, we can simply recreate them in anyway we want with routing action: `tunnel_from_encap`. It allows the source and target encap and their override value being set, which allows us to preserve the encaps in anyway we want and also ensures the clarity of final transformation that we are doing in the end.
```

Comment:
```suggestion
Since all of the encap information is preserved in the metadata bus for flow creation, to solve this problem, we can simply recreate them in the manner we want using the routing action: `tunnel_from_encap`. It allows the source and target encap and their override value to be set, which allows us to preserve the encaps exactly as intended, and also ensures the clarity of final transformation performed in the end.
```